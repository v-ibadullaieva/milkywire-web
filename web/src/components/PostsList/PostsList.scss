.List {
  .ant-list-items {
    display: flex;
    flex-wrap: wrap;
    justify-content: space-between;
  }
  .ant-card-cover img {
    height: 160px;
    object-fit: cover;
    
    &:hover {
      cursor: pointer;
    }
  }
  .ant-card-extra {
    width: 100%;

    & > div {
      display: flex;
      justify-content: space-between;
      align-items: center;
    }
  }
}
