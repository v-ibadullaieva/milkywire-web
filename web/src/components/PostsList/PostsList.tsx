import React, { useEffect, useState } from 'react';
import { Card, List, Modal, Icon } from 'antd';

import './PostsList.scss';
import { fetchPosts, deletePost, editPost } from '../../api/posts';
import Post from '../Post/Post';
import EditForm from '../EditForm/EditForm';
import { UpdatedPostType } from '../EditForm/EditForm';

export interface PostType extends UpdatedPostType {
  id: string
};

interface MoreProps extends React.Props<any> {
  post: PostType,
  handleDelete: (id: string) => void,
  handleOpenForm: (id: string) => void
}

const { Meta } = Card;

const PostsList = () => {
  const [posts, setPosts] = useState<PostType[]>([]);
  const [visible, setVisible] = useState<boolean>(false);
  const [visibleForm, setVisibleForm] = useState<boolean>(false);
  const [chosen, setChosen] = useState<string | null>(null);

  const getPosts = async () => {
    const posts = await fetchPosts();
    setPosts(posts);
  }

  useEffect(() => {
    getPosts();
  }, []);

  const handleOpen = (id: string | null) => {
    setChosen(id);
    setVisible(true);
  }
  
  const handleOpenForm = (id: string | null) => {
    setChosen(id);
    setVisibleForm(true);
  }

  const handleDeletePost = (id: string) => {
    deletePost(id).then(getPosts);
  }

  const handleEditPost = (data: PostType) => {
    editPost(data).then(getPosts);
  }

  const handleNext = () => {
    if (chosenIndex >= posts.length - 1) return;
    setChosen(posts[chosenIndex+1].id)
  }

  const handlePrev = () => {
    if (chosenIndex === 0) return;
    setChosen(posts[chosenIndex-1].id)
  }

  const chosenIndex = posts.findIndex(post => post.id === chosen);
  const chosenPost = posts[chosenIndex];

  return (
    <div>
      <List
        className='List'
        itemLayout='vertical'
        size='large'
        pagination={{
          pageSize: 15
        }}
        dataSource={posts}
        renderItem={post => (
          <List.Item>
            <Card
              key={post.id}
              style={{ width: 240 }}
              cover={
                <img alt='img' src={post.data.media[0].image} onClick={() => handleOpen(post.id)} className='cardImg' />
              }
              extra={<More post={post} handleDelete={handleDeletePost} handleOpenForm={handleOpenForm} />}
            >
              <Meta title={post.data.media[0].description} description={post.description} />
            </Card>
          </List.Item>
        )}/>
        <Modal
          title={chosen ? chosenPost.data.media[0].description : 'Post'}
          visible={visible}
          footer={null}
          onCancel={() => setVisible(false)}
        >
          <Post post={chosenPost} handleNext={handleNext} handlePrev={handlePrev} />
        </Modal>
        <Modal
          title={chosen ? chosenPost.data.media[0].description : 'Post'}
          visible={visibleForm}
          footer={null}
          onCancel={() => setVisibleForm(false)}
        >
          <EditForm post={chosenPost} onCancel={() => setVisibleForm(false)} editPost={handleEditPost} />
        </Modal>
      </div>
  )
}

const More = ({ post, handleDelete, handleOpenForm }: MoreProps) => (
  <div>
    <a onClick={() => handleOpenForm(post.id)}>Edit</a>
    <Icon type='delete' theme='twoTone' onClick={() => handleDelete(post.id)} />
  </div>
);

export default PostsList;