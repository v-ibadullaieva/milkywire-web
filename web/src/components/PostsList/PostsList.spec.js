import React from 'react';
import { create } from "react-test-renderer";
import { act } from "react-dom/test-utils";
import { render, unmountComponentAtNode } from "react-dom";

import PostsList from './PostsList';

const POSTS = [
  {
    id: 0,
    type: "IMAGES",
    description: "lol",
    data: {
      media: [{
        image: "https://picsum.photos/id/0/5616/3744",
        width: 5616,
        height: 3744,
        version: "2019-03-14",
        description: "Alejandro Escamilla"
      }]
    },
    impacter_id: 0
  },
  {
    id: 1,
    type: "IMAGES",
    description: "sint excepteur irure ut amet adipisicing Lorem aliqua adipisicing amet",
    data: {
      media: [{
        image: "https://picsum.photos/id/0/5616/3744",
        width: 5616,
        height: 3744,
        version: "2019-03-14",
        description: "Paul Jarvis"
      }]
    },
    impacter_id: 0
  }
];
let container;

beforeEach(() => {
  container = document.createElement("div");
  document.body.appendChild(container);
});

afterEach(() => {
  unmountComponentAtNode(container);
  container.remove();
  container = null;
});

describe('PostsList component', () => {
  test("Matches the snapshot", () => {
    const postsList = create(<PostsList />);
    expect(postsList.toJSON()).toMatchSnapshot();
  });

  test("it shows a list of posts", async () => {
    jest.spyOn(window, "fetch").mockImplementation(() => {
      const fetchResponse = {
        json: () => Promise.resolve(POSTS)
      };
      return Promise.resolve(fetchResponse);
    });

    await act(async () => {
      render(<PostsList />, container);
    });

    expect(container.textContent).toBe("EditAlejandro EscamillalolEditPaul Jarvissint excepteur irure ut amet adipisicing Lorem aliqua adipisicing amet1");
    window.fetch.mockRestore();
  });
});
