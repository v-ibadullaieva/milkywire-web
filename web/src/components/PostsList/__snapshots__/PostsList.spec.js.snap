// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PostsList component Matches the snapshot 1`] = `
<div>
  <div
    className="ant-list List ant-list-vertical ant-list-lg ant-list-split ant-list-something-after-last-item"
  >
    <div
      className="ant-spin-nested-loading"
    >
      <div
        className="ant-spin-container"
      >
        <div
          className="ant-list-empty-text"
        >
          <div
            className="ant-empty ant-empty-normal"
          >
            <div
              className="ant-empty-image"
            >
              <svg
                height="41"
                viewBox="0 0 64 41"
                width="64"
                xmlns="http://www.w3.org/2000/svg"
              >
                <g
                  fill="none"
                  fillRule="evenodd"
                  transform="translate(0 1)"
                >
                  <ellipse
                    cx="32"
                    cy="33"
                    fill="#F5F5F5"
                    rx="32"
                    ry="7"
                  />
                  <g
                    fillRule="nonzero"
                    stroke="#D9D9D9"
                  >
                    <path
                      d="M55 12.76L44.854 1.258C44.367.474 43.656 0 42.907 0H21.093c-.749 0-1.46.474-1.947 1.257L9 12.761V22h46v-9.24z"
                    />
                    <path
                      d="M41.613 15.931c0-1.605.994-2.93 2.227-2.931H55v18.137C55 33.26 53.68 35 52.05 35h-40.1C10.32 35 9 33.259 9 31.137V13h11.16c1.233 0 2.227 1.323 2.227 2.928v.022c0 1.605 1.005 2.901 2.237 2.901h14.752c1.232 0 2.237-1.308 2.237-2.913v-.007z"
                      fill="#FAFAFA"
                    />
                  </g>
                </g>
              </svg>
            </div>
            <p
              className="ant-empty-description"
            >
              No Data
            </p>
          </div>
        </div>
      </div>
    </div>
    <div
      className="ant-list-pagination"
    >
      <ul
        className="ant-pagination"
        style={Object {}}
        unselectable="unselectable"
      >
        <li
          aria-disabled={true}
          className="ant-pagination-disabled ant-pagination-prev"
          onClick={[Function]}
          onKeyPress={[Function]}
          tabIndex={null}
          title="Previous Page"
        >
          <a
            className="ant-pagination-item-link"
          >
            <i
              aria-label="icon: left"
              className="anticon anticon-left"
            >
              <svg
                aria-hidden="true"
                className=""
                data-icon="left"
                fill="currentColor"
                focusable="false"
                height="1em"
                viewBox="64 64 896 896"
                width="1em"
              >
                <path
                  d="M724 218.3V141c0-6.7-7.7-10.4-12.9-6.3L260.3 486.8a31.86 31.86 0 0 0 0 50.3l450.8 352.1c5.3 4.1 12.9.4 12.9-6.3v-77.3c0-4.9-2.3-9.6-6.1-12.6l-360-281 360-281.1c3.8-3 6.1-7.7 6.1-12.6z"
                />
              </svg>
            </i>
          </a>
        </li>
        <li
          className="ant-pagination-item ant-pagination-item-0 ant-pagination-disabled ant-pagination-item-disabled"
          onClick={[Function]}
          onKeyPress={[Function]}
          tabIndex="0"
          title={0}
        >
          <a>
            0
          </a>
        </li>
        <li
          aria-disabled={true}
          className="ant-pagination-disabled ant-pagination-next"
          onClick={[Function]}
          onKeyPress={[Function]}
          tabIndex={null}
          title="Next Page"
        >
          <a
            className="ant-pagination-item-link"
          >
            <i
              aria-label="icon: right"
              className="anticon anticon-right"
            >
              <svg
                aria-hidden="true"
                className=""
                data-icon="right"
                fill="currentColor"
                focusable="false"
                height="1em"
                viewBox="64 64 896 896"
                width="1em"
              >
                <path
                  d="M765.7 486.8L314.9 134.7A7.97 7.97 0 0 0 302 141v77.3c0 4.9 2.3 9.6 6.1 12.6l360 281.1-360 281.1c-3.9 3-6.1 7.7-6.1 12.6V883c0 6.7 7.7 10.4 12.9 6.3l450.8-352.1a31.96 31.96 0 0 0 0-50.4z"
                />
              </svg>
            </i>
          </a>
        </li>
      </ul>
    </div>
  </div>
</div>
`;
